File: Solution O(n) Space.java

import java.util.ArrayDeque;
import java.util.Deque;

public class Solution {

    public int longestValidParentheses(String input) {
        Deque<Integer> stackIndexes = new ArrayDeque<>();
        final int sentinel = -1;
        stackIndexes.push(sentinel);
        int maxValidLength = 0;

        for (int i = 0; i < input.length(); ++i) {
            if (input.charAt(i) == '(') {
                stackIndexes.push(i);
                continue;
            }
            stackIndexes.pop();
            if (stackIndexes.isEmpty()) {
                stackIndexes.push(i);
                continue;
            }
            maxValidLength = Math.max(maxValidLength, i - stackIndexes.peek());
        }
        return maxValidLength;
    }
}

File: Solution O(n) Space.js

/**
 * @param {string} input
 * @return {number}
 */
var longestValidParentheses = function (input) {
    const stackIndexes = [];
    const sentinel = -1;
    stackIndexes.push(sentinel);
    let maxValidLength = 0;

    for (let i = 0; i < input.length; ++i) {
        if (input.charAt(i) === '(') {
            stackIndexes.push(i);
            continue;
        }
        stackIndexes.pop();
        if (stackIndexes.length === 0) {
            stackIndexes.push(i);
            continue;
        }
        maxValidLength = Math.max(maxValidLength, i - stackIndexes[stackIndexes.length - 1]);
    }
    return maxValidLength;
};
